File: java/src/main/java/io/cucumber/query/Repository.java
package io.cucumber.query;

import io.cucumber.messages.types.Attachment;
import io.cucumber.messages.types.Envelope;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Hook;
import io.cucumber.messages.types.Meta;
import io.cucumber.messages.types.Pickle;
import io.cucumber.messages.types.PickleStep;
import io.cucumber.messages.types.Scenario;
import io.cucumber.messages.types.Step;
import io.cucumber.messages.types.StepDefinition;
import io.cucumber.messages.types.Suggestion;
import io.cucumber.messages.types.TestCase;
import io.cucumber.messages.types.TestCaseFinished;
import io.cucumber.messages.types.TestCaseStarted;
import io.cucumber.messages.types.TestRunFinished;
import io.cucumber.messages.types.TestRunStarted;
import io.cucumber.messages.types.TestStep;
import io.cucumber.messages.types.TestStepFinished;
import io.cucumber.messages.types.TestStepStarted;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.BiFunction;

import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_ATTACHMENTS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_GHERKIN_DOCUMENTS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_HOOKS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_STEP_DEFINITIONS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_SUGGESTIONS;

/**
 * A write only repository of Cucumber Messages.
 * <p>
 * This class is effectively a simple in memory database. It can be updated in
 * through the {@link #update(Envelope)} method, and be queried by {@link Query}.
 */
public final class Repository {
    private final Set<RepositoryFeature> features;

    final Map<String, TestCaseStarted> testCaseStartedById = new LinkedHashMap<>();
    final Map<String, TestCaseFinished> testCaseFinishedByTestCaseStartedId = new LinkedHashMap<>();
    final Map<String, List<TestStepFinished>> testStepsFinishedByTestCaseStartedId = new LinkedHashMap<>();
    final Map<String, List<TestStepStarted>> testStepsStartedByTestCaseStartedId = new LinkedHashMap<>();
    final Map<String, Pickle> pickleById = new LinkedHashMap<>();
    final Map<String, TestCase> testCaseById = new LinkedHashMap<>();
    final Map<String, Step> stepById = new LinkedHashMap<>();
    final Map<String, TestStep> testStepById = new LinkedHashMap<>();
    final Map<String, PickleStep> pickleStepById = new LinkedHashMap<>();
    final Map<String, Hook> hookById = new LinkedHashMap<>();
    final Map<String, List<Attachment>> attachmentsByTestCaseStartedId = new LinkedHashMap<>();
    final Map<Object, Lineage> lineageById = new HashMap<>();
    final Map<String, StepDefinition> stepDefinitionById = new LinkedHashMap<>();
    final Map<String, List<Suggestion>> suggestionsByPickleStepId = new LinkedHashMap<>();

    Meta meta;
    TestRunStarted testRunStarted;
    TestRunFinished testRunFinished;

    private Repository(Set<RepositoryFeature> features) {
        this.features = features;
    }

    public static Builder builder() {
        return new Builder();
    }

    public enum RepositoryFeature {

        /**
         * Include {@link Attachment} messages.
         * <p>
         * Disable to reduce memory usage.
         */
        INCLUDE_ATTACHMENTS,

        /**
         * Include {@link io.cucumber.messages.types.GherkinDocument} messages.
         * <p>
         * Disable to reduce memory usage.
         */
        INCLUDE_GHERKIN_DOCUMENTS,

        /**
         * Include {@link Hook} messages.
         * <p>
         * Disable to reduce memory usage.
         */
        INCLUDE_HOOKS,

        /**
         * Include {@link StepDefinition} messages.
         * <p>
         * Disable to reduce memory usage.
         */
        INCLUDE_STEP_DEFINITIONS,

        /**
         * Include {@link Suggestion} messages.
         * <p>
         * Disable to reduce memory usage.
         */
        INCLUDE_SUGGESTIONS,
    }

    public static class Builder {
        private final EnumSet<RepositoryFeature> features = EnumSet.noneOf(RepositoryFeature.class);
        
        private Builder(){
            
        }
        /**
         * Toggles a given feature.
         */
        public Builder feature(RepositoryFeature feature, boolean enabled) {
            if (enabled) {
                features.add(feature);
            } else {
                features.remove(feature);
            }
            return this;
        }

        public Repository build() {
            return new Repository(features);
        }
    }

    public void update(Envelope envelope) {
        envelope.getMeta().ifPresent(this::updateMeta);
        envelope.getTestRunStarted().ifPresent(this::updateTestRunStarted);
        envelope.getTestRunFinished().ifPresent(this::updateTestRunFinished);
        envelope.getTestCaseStarted().ifPresent(this::updateTestCaseStarted);
        envelope.getTestCaseFinished().ifPresent(this::updateTestCaseFinished);
        envelope.getTestStepStarted().ifPresent(this::updateTestStepStarted);
        envelope.getTestStepFinished().ifPresent(this::updateTestStepFinished);
        envelope.getPickle().ifPresent(this::updatePickle);
        envelope.getTestCase().ifPresent(this::updateTestCase);

        if (features.contains(INCLUDE_GHERKIN_DOCUMENTS)) {
            envelope.getGherkinDocument().ifPresent(this::updateGherkinDocument);
        }
        if (features.contains(INCLUDE_STEP_DEFINITIONS)) {
            envelope.getStepDefinition().ifPresent(this::updateStepDefinition);
        }
        if (features.contains(INCLUDE_HOOKS)) {
            envelope.getHook().ifPresent(this::updateHook);
        }
        if (features.contains(INCLUDE_ATTACHMENTS)) {
            envelope.getAttachment().ifPresent(this::updateAttachment);
        }
        if (features.contains(INCLUDE_SUGGESTIONS)) {
            envelope.getSuggestion().ifPresent(this::updateSuggestions);
        }
    }

    private void updateAttachment(Attachment attachment) {
        attachment.getTestCaseStartedId()
                .ifPresent(testCaseStartedId -> this.attachmentsByTestCaseStartedId.compute(testCaseStartedId, updateList(attachment)));
    }

    private void updateHook(Hook hook) {
        this.hookById.put(hook.getId(), hook);
    }

    private void updateTestCaseStarted(TestCaseStarted testCaseStarted) {
        this.testCaseStartedById.put(testCaseStarted.getId(), testCaseStarted);
    }

    private void updateTestCase(TestCase event) {
        this.testCaseById.put(event.getId(), event);
        event.getTestSteps().forEach(testStep -> testStepById.put(testStep.getId(), testStep));
    }

    private void updatePickle(Pickle event) {
        this.pickleById.put(event.getId(), event);
        event.getSteps().forEach(pickleStep -> pickleStepById.put(pickleStep.getId(), pickleStep));
    }

    private void updateGherkinDocument(GherkinDocument document) {
        lineageById.putAll(Lineages.of(document));
        document.getFeature().ifPresent(this::updateFeature);
    }

    private void updateFeature(io.cucumber.messages.types.Feature feature) {
        feature.getChildren()
                .forEach(featureChild -> {
                    featureChild.getBackground().ifPresent(background -> updateSteps(background.getSteps()));
                    featureChild.getScenario().ifPresent(this::updateScenario);
                    featureChild.getRule().ifPresent(rule -> rule.getChildren().forEach(ruleChild -> {
                        ruleChild.getBackground().ifPresent(background -> updateSteps(background.getSteps()));
                        ruleChild.getScenario().ifPresent(this::updateScenario);
                    }));
                });
    }

    private void updateTestStepStarted(TestStepStarted event) {
        this.testStepsStartedByTestCaseStartedId.compute(event.getTestCaseStartedId(), updateList(event));
    }

    private void updateTestStepFinished(TestStepFinished event) {
        this.testStepsFinishedByTestCaseStartedId.compute(event.getTestCaseStartedId(), updateList(event));
    }

    private void updateTestCaseFinished(TestCaseFinished event) {
        this.testCaseFinishedByTestCaseStartedId.put(event.getTestCaseStartedId(), event);
    }

    private void updateTestRunFinished(TestRunFinished event) {
        this.testRunFinished = event;
    }

    private void updateTestRunStarted(TestRunStarted event) {
        this.testRunStarted = event;
    }

    private void updateScenario(Scenario scenario) {
        updateSteps(scenario.getSteps());
    }

    private void updateStepDefinition(StepDefinition event) {
        this.stepDefinitionById.put(event.getId(), event);
    }

    private void updateSteps(List<Step> steps) {
        steps.forEach(step -> stepById.put(step.getId(), step));
    }

    private void updateSuggestions(Suggestion event) {
        this.suggestionsByPickleStepId.compute(event.getPickleStepId(), updateList(event));
    }

    private void updateMeta(Meta event) {
        this.meta = event;
    }

    private <K, E> BiFunction<K, List<E>, List<E>> updateList(E element) {
        return (key, existing) -> {
            if (existing != null) {
                existing.add(element);
                return existing;
            }
            List<E> list = new ArrayList<>();
            list.add(element);
            return list;
        };
    }

}

File: dotnet/Query/Query/Repository.cs
#nullable enable
using Io.Cucumber.Messages.Types;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System;
using System.Linq;

namespace Io.Cucumber.Query
{
    public class Repository
    {
        // Features
        public enum RepositoryFeature
        {
            INCLUDE_ATTACHMENTS,
            INCLUDE_GHERKIN_DOCUMENTS,
            INCLUDE_HOOKS,
            INCLUDE_STEP_DEFINITIONS,
            INCLUDE_SUGGESTIONS,
        }

        private readonly HashSet<RepositoryFeature> _features;

        // Public fields (matching Java order)
        public readonly Dictionary<string, TestCaseStarted> TestCaseStartedById = new();
        public readonly Dictionary<string, TestCaseFinished> TestCaseFinishedByTestCaseStartedId = new();
        public readonly Dictionary<string, List<TestStepFinished>> TestStepsFinishedByTestCaseStartedId = new();
        public readonly Dictionary<string, List<TestStepStarted>> TestStepsStartedByTestCaseStartedId = new();
        public readonly Dictionary<string, TestRunHookStarted> TestRunHookStartedById = new();
        public readonly Dictionary<string, TestRunHookFinished> TestRunHookFinishedByTestRunHookStartedId = new();
        public readonly Dictionary<string, Pickle> PickleById = new();
        public readonly Dictionary<string, TestCase> TestCaseById = new();
        public readonly Dictionary<string, Step> StepById = new();
        public readonly Dictionary<string, TestStep> TestStepById = new();
        public readonly Dictionary<string, PickleStep> PickleStepById = new();
        public readonly Dictionary<string, Hook> HookById = new();
        public readonly Dictionary<string, List<Attachment>> AttachmentsByTestCaseStartedId = new();
        public readonly Dictionary<string, List<Attachment>> AttachmentsByTestRunHookStartedId = new();
        public readonly Dictionary<object, Lineage> LineageById = new();
        public readonly Dictionary<string, StepDefinition> StepDefinitionById = new();
        public readonly Dictionary<string, List<Suggestion>> SuggestionsByPickleStepId = new();

        public Meta? Meta;
        public TestRunStarted? TestRunStarted;
        public TestRunFinished? TestRunFinished;

        private Repository(HashSet<RepositoryFeature> features)
        {
            _features = features;
        }

        public static RepositoryBuilder Builder() => new RepositoryBuilder();

        public class RepositoryBuilder
        {
            private readonly HashSet<RepositoryFeature> _features = new();
            public RepositoryBuilder Feature(RepositoryFeature feature, bool enabled)
            {
                if (enabled) _features.Add(feature);
                else _features.Remove(feature);
                return this;
            }
            public Repository Build() => new Repository(new HashSet<RepositoryFeature>(_features));
        }

        public void Update(Envelope envelope)
        {
            if (envelope.Meta != null) UpdateMeta(envelope.Meta);
            if (envelope.TestRunStarted != null) UpdateTestRunStarted(envelope.TestRunStarted);
            if (envelope.TestRunFinished != null) UpdateTestRunFinished(envelope.TestRunFinished);
            if (envelope.TestRunHookStarted != null) UpdateTestRunHookStarted(envelope.TestRunHookStarted);
            if (envelope.TestRunHookFinished != null) UpdateTestRunHookFinished(envelope.TestRunHookFinished);
            if (envelope.TestCaseStarted != null) UpdateTestCaseStarted(envelope.TestCaseStarted);
            if (envelope.TestCaseFinished != null) UpdateTestCaseFinished(envelope.TestCaseFinished);
            if (envelope.TestStepStarted != null) UpdateTestStepStarted(envelope.TestStepStarted);
            if (envelope.TestStepFinished != null) UpdateTestStepFinished(envelope.TestStepFinished);
            if (envelope.Pickle != null) UpdatePickle(envelope.Pickle);
            if (envelope.TestCase != null) UpdateTestCase(envelope.TestCase);
            if (_features.Contains(RepositoryFeature.INCLUDE_GHERKIN_DOCUMENTS) && envelope.GherkinDocument != null) UpdateGherkinDocument(envelope.GherkinDocument);
            if (_features.Contains(RepositoryFeature.INCLUDE_STEP_DEFINITIONS) && envelope.StepDefinition != null) UpdateStepDefinition(envelope.StepDefinition);
            if (_features.Contains(RepositoryFeature.INCLUDE_HOOKS) && envelope.Hook != null) UpdateHook(envelope.Hook);
            if (_features.Contains(RepositoryFeature.INCLUDE_ATTACHMENTS) && envelope.Attachment != null) UpdateAttachment(envelope.Attachment);
            if (_features.Contains(RepositoryFeature.INCLUDE_SUGGESTIONS) && envelope.Suggestion != null) UpdateSuggestions(envelope.Suggestion);
        }

        public void UpdateAttachment(Attachment attachment)
        {
            if (attachment.TestCaseStartedId != null)
            {
                if (!AttachmentsByTestCaseStartedId.TryGetValue(attachment.TestCaseStartedId, out var list))
                {
                    list = new List<Attachment>();
                    AttachmentsByTestCaseStartedId[attachment.TestCaseStartedId] = list;
                }
                list.Add(attachment);
            }
            if (attachment.TestRunHookStartedId != null)
            {
                if (!AttachmentsByTestRunHookStartedId.TryGetValue(attachment.TestRunHookStartedId, out var list))
                {
                    list = new List<Attachment>();
                    AttachmentsByTestRunHookStartedId[attachment.TestRunHookStartedId] = list;
                }
                list.Add(attachment);
            }
        }

        public void UpdateHook(Hook hook)
        {
            HookById[hook.Id] = hook;
        }

        public void UpdateTestCaseStarted(TestCaseStarted testCaseStarted)
        {
            TestCaseStartedById[testCaseStarted.Id] = testCaseStarted;
        }

        public void UpdateTestCase(TestCase testCase)
        {
            TestCaseById[testCase.Id] = testCase;
            foreach (var testStep in testCase.TestSteps)
            {
                TestStepById[testStep.Id] = testStep;
            }
        }

        public void UpdatePickle(Pickle pickle)
        {
            PickleById[pickle.Id] = pickle;
            foreach (var step in pickle.Steps)
            {
                PickleStepById[step.Id] = step;
            }
        }

        public void UpdateGherkinDocument(GherkinDocument document)
        {
            foreach (var lineage in Lineages.Of(document))
            {
                LineageById.Add(lineage.Key, lineage.Value);
            }
            if (document.Feature != null)
            {
                UpdateFeature(document.Feature);
            }
        }

        public void UpdateFeature(Feature feature)
        {
            foreach (var child in feature.Children)
            {
                if (child.Background != null)
                {
                    UpdateSteps(child.Background.Steps);
                }
                if (child.Scenario != null)
                {
                    UpdateScenario(child.Scenario);
                }
                if (child.Rule != null)
                {
                    foreach (var ruleChild in child.Rule.Children)
                    {
                        if (ruleChild.Background != null)
                        {
                            UpdateSteps(ruleChild.Background.Steps);
                        }
                        if (ruleChild.Scenario != null)
                        {
                            UpdateScenario(ruleChild.Scenario);
                        }
                    }
                }
            }
        }

        public void UpdateTestStepStarted(TestStepStarted testStepStarted)
        {
            if (!TestStepsStartedByTestCaseStartedId.TryGetValue(testStepStarted.TestCaseStartedId, out var list))
            {
                list = new List<TestStepStarted>();
                TestStepsStartedByTestCaseStartedId[testStepStarted.TestCaseStartedId] = list;
            }
            list.Add(testStepStarted);
        }

        public void UpdateTestStepFinished(TestStepFinished testStepFinished)
        {
            if (!TestStepsFinishedByTestCaseStartedId.TryGetValue(testStepFinished.TestCaseStartedId, out var list))
            {
                list = new List<TestStepFinished>();
                TestStepsFinishedByTestCaseStartedId[testStepFinished.TestCaseStartedId] = list;
            }
            list.Add(testStepFinished);
        }

        public void UpdateTestCaseFinished(TestCaseFinished testCaseFinished)
        {
            TestCaseFinishedByTestCaseStartedId[testCaseFinished.TestCaseStartedId] = testCaseFinished;
        }

        public void UpdateTestRunFinished(TestRunFinished testRunFinished)
        {
            TestRunFinished = testRunFinished;
        }

        public void UpdateTestRunStarted(TestRunStarted testRunStarted)
        {
            TestRunStarted = testRunStarted;
        }

        public void UpdateTestRunHookStarted(TestRunHookStarted testRunHookStarted)
        {
            TestRunHookStartedById[testRunHookStarted.Id] = testRunHookStarted;
        }

        public void UpdateTestRunHookFinished(TestRunHookFinished testRunHookFinished)
        {
            TestRunHookFinishedByTestRunHookStartedId[testRunHookFinished.TestRunHookStartedId] = testRunHookFinished;
        }

        public void UpdateScenario(Scenario scenario)
        {
            UpdateSteps(scenario.Steps);
        }

        public void UpdateStepDefinition(StepDefinition stepDefinition)
        {
            StepDefinitionById[stepDefinition.Id] = stepDefinition;
        }

        public void UpdateSteps(IList<Step> steps)
        {
            foreach (var step in steps)
            {
                StepById[step.Id] = step;
            }
        }

        public void UpdateSuggestions(Suggestion suggestion)
        {
            if (!SuggestionsByPickleStepId.TryGetValue(suggestion.PickleStepId, out var list))
            {
                list = new List<Suggestion>();
                SuggestionsByPickleStepId[suggestion.PickleStepId] = list;
            }
            list.Add(suggestion);
        }

        public void UpdateMeta(Meta meta)
        {
            Meta = meta;
        }
    }
}